### DISCUSSION TITLE
Recording paths of rays with PTracer Integrator and finding out how many hit a surface

### DISCUSSION THREAD

**Discussion by h-OUS-e** [mitsuba-renderer/mitsuba3#559] [Q&A] [answered]
Hello,

Sorry for the many questions, this topic is new to me. I would like to sample my own rays, and use these rays with the PTracer Integrator, and then record their paths. This is just to help visualize a few rays and how they travel through my scene setup. Is that possible without making my own integrator? The integrator examples I found written in python usually start from a sensor, but I was searching for something like a PTracer where rays are emitted from some location. 

I would also like to calculate how many rays hit a particular spot on a mesh object to say that the more rays hit that same spot, the better it is, to guide an optimization problem.

Many thanks!

**Comment by njroussel** [comment, marked as answer]
> I would like to sample my own rays, and use these rays with the PTracer Integrator, and then record their paths.

This is not possible out-of-the box. You would need to write your own integrator or at least replicate the main loop of the `ptracer`'s `sample` method in a script.

>I would also like to calculate how many rays hit a particular spot on a mesh object to say that the more rays hit that same spot, the better it is, to guide an optimization problem.

If you're only working with meshes, the [`SurfaceInteraction3f`](https://mitsuba.readthedocs.io/en/stable/src/api_reference.html#mitsuba.SurfaceInteraction3f) object (resulting of a ray-shape intersection) holds a pointer to the shape. If you need something more precise you can use the `uv` coordinates (assuming your object has a UV-mapping) or the `prim_index` field holds the index of the triangle the ray hit.

**Reply by h-OUS-e** [reply]
Thank you @njroussel, as a follow-up, is the ptracer's sample method written in Python? I couldn't find it in the integrator python example. Would I need to write one in C ++?

For the surfaceInteraction3f, how would I get that object from the scene after I render the scene with `mi.render()`?

**Reply by njroussel** [reply]
No it is [written in C++](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/integrators/ptracer.cpp), but you can port/rewrite it in Python.

You can not access it at the end of your render, you need access it at every step (bounce/intersection) of your integrator/script. Take another look at [this tutorial](https://mitsuba.readthedocs.io/en/stable/src/rendering/scripting_renderer.html) how we get it from a ray intersection.

**Reply by h-OUS-e** [reply]
I am trying to write a Python version of the PTracer. From what I understoor, I can't inherit `mi.SamplingIntegrator` as it records rays starting from the sensor. Instead, I should inherit from `mi.AdJointIntegrator` since it does the rendering for rays starting from the light source. However, when I try to write:

```
lass MyPtracer(mi.AdjointIntegrator):

    def __init__(self, props=mi.Properties()):
        super().__init__(props)
        print("test")
```
and then register the integrator with `mi.register_integrator("MyPtracer", lambda props: MyPtracer(props))`
I get the error TypeError: MyPtracer: No constructor defined! after loading the scene. This error doesn't appear when I inherit from mi.SamplingIntegrator. How else should I write the initiation of the PTracer in python?

Sorry if this question is redundant, this is all a new territory to me.

**Reply by njroussel** [reply]
I believe this is a bug/limitation on our end. We have not enable the mechanism to extend `AdjointIntegrator` types. I will add that soon. I'll keep this thread updated.

**Reply by njroussel** [reply]
https://github.com/mitsuba-renderer/mitsuba3/commit/c4a8b31ee764a0e6d56d9075708c3c76062854be

This commit added the mechanism to extend `AdjointIntegrator`.
